lib/elixir/lib/enum.ex

```diff
@@ -2855,7 +2857,7 @@ defimpl Enumerable, for: Map do
     {:ok, false}
   end
 
-  def reduce(map, acc, fun) do
+  def reduce(map, acc, fun) when is_function(fun, 2) do
```

Comment:
I would maybe skip this as we do the checks everywhere else now so unless this is called directly (which it usually isn't) it's gonna get the right function with arity 2. Same goes for `Enumerable.List` implementation. Wdyt?